### ISSUE TITLE
Question/Suggestion: Are the masks used for pruning soft or hard?

### ISSUE THREAD

**Issue by MahdiNazemi** [NervanaSystems/distiller#373]
Based on my understanding of the code, `zeros_mask_dict` keeps a mask for the pruned tensors. Is this a soft mask or a hard mask?

For example, when using AGP, some weights are set to zero during each iteration. Are these weights allowed to become nonzero later or will remain zero?

Is there an option that allows choosing between a soft and hard mask?

**Comment by nzmora** [comment]
Hi @MahdiNazemi 

The masks are hard, meaning that they have discrete binary values (0/1).  The term "soft masks", as I am familiar with it, means that the masks are real-valued in the range [0..1].  

So I interpret your question not as asking if Distiller uses soft masks, but if the masks remain unchanged once set.  
This is a property of the pruner you use.  For example, the "vanilla" AGP pruner uses a pruning schedule to determine *how much* to prune at each step (pruning regimen), and uses L1-magnitude to decide *which* connections to prune (pruning criterion).  If instead if using L1-magnitude you decide to use some other pruning criterion that prunes different connections each time the mask is set, then connections will be revived.  An extreme (and not very useful) example is using AGP with a random pruner.  AGP controls how much we prune, and the random criterion chooses which connections to prune.

We have implemented a [`SplicingPruner`](https://nervanasystems.github.io/distiller/algo_pruning.html#splicing-pruner) that uses L1-magnitude and also allows weights in a certain region around the decision boundary to "return to life".  You can find example YAML schedules [here](https://github.com/NervanaSystems/distiller/tree/master/examples/network_surgery).  I personally don't like it, because it has so many hyper-parameters.  Maybe one day we'll change our `SplicingPruner` [implementation](https://github.com/NervanaSystems/distiller/blob/master/distiller/pruning/splicing_pruner.py) to use sparsity percentages for the "network surgery function" instead of using "sensitivity" values.  That will reduce many hyperparameters.

We allow for [fine-control over the pruning process](https://nervanasystems.github.io/distiller/schedule.html#pruning-fine-control), so you create more complex pruning schemes.

Cheers
Neta

**Comment by hunterkun** [comment]
Hi, @MahdiNazemi , in my understanding, the above you said is the default way that pruning happens, that is the weights that are set to zero in one iteration remain zero in later iteration and never revive again. This is because although the weights would get updated , they would be masked again as the attached `masker` in the end of mini-batch by `CompressionScheduler`. 

However, as the @nzmora  said, it has provided more fine control operation.  For example, if you want the zero weights "revive" again, you can add `mask_on_forward_only` option and you can even decide set changing frequency by `mini_batch_pruning_frequency`.

If i have some misunderstanding, feel free to point out. @nzmora. Hope it would help to @MahdiNazemi

**Comment by nzmora** [comment]
Hi @MahdiNazemi,
> In vanilla AGP, the weights that are set to zero in one iteration remain zero throughout the process and some other non-zero weights are added during later iterations. Using your terminology, the weights that are set to zero in earlier iterations, don't return to life in later iterations. Is that true?

Yes, this is correct.  Thanks @hunterkun 
Cheers,
Neta
